### PR TITLE
Task/2327 database pooling

### PR DESCRIPTION
**Pull request by maximiliancsuk**

### CODE REVIEW

File: cmd/metrics-receiver/main.go [hunk 1]
```diff
@@ -21,7 +21,7 @@ import (
 
 var (
 	version    = "0.0.0-src"
-	configFile = flag.String("config", "config.json", "Config file location")
+	configFile = flag.String("config", "C:\\Users\\MuhametKaçuri\\repos\\metrics-receiver-ng\\config-example.json", "Config file location")
```
Comment (maximiliancsuk) [inline cmd/metrics-receiver/main.go]: Please remove that, it's not ideal to have local paths in checked-in source code

File: pkg/timescale/write.go [hunk 2]
```diff
@@ -1,15 +1,50 @@
...
+
+	for _, output := range cfg {
+
+		c, errC := pgxpool.ParseConfig(output.Connection)
+
+		if errC != nil {
+			return fmt.Errorf("Unable to parse config for timescale database: %v\n", errC)
+		}
+
+		timescaleDbPool, err := pgxpool.ConnectConfig(context.Background(), c)
```
Comment (maximiliancsuk) [inline pkg/timescale/write.go:29]: Does this already connect to the database? What happens when (at least one of) the databases are not reachable on init?

In general, can the pool recover when a database connection disconnects during runtime, but then later becomes reachable again?

File: pkg/timescale/write.go [hunk 3]
```diff
@@ -1,15 +1,50 @@
...
+		}
+
+		timescalePools[output.Connection] = timescaleDbPool
+
+	}
+
+	return nil
+}
+
+func CloseConnectionPools() {
```
Comment (maximiliancsuk) [inline pkg/timescale/write.go:42]: You are never calling this?

File: pkg/timescale/write.go [hunk 4]
```diff
@@ -112,39 +147,37 @@ func contains(s []string, e string) bool {
...
 		}
 	}
 
-	err := tx.Commit()
+	err := tx.Commit(context.Background())
 	if err != nil {
 		return err
 	}
 
+	conn.Release()
```
Comment (maximiliancsuk) [inline pkg/timescale/write.go]: That's a bug... if the function returns early because of an error, the connection is never relased... work with `defer` instead?

File: pkg/timescale/write.go [hunk 5]
```diff
@@ -112,39 +147,37 @@ func contains(s []string, e string) bool {
 }
 
 func insertRowsTimescale(rowsArray []TimescaleRows, config *config.OutputTimescale) error {
-	var c, parseErr = pgx.ParseConnectionString(config.Connection)
-	if parseErr != nil {
-		return parseErr
-	}
 
-	conn, connErr := pgx.Connect(c)
+	conn, connErr := timescalePools[config.Connection].Acquire(context.Background())
```
Comment (maximiliancsuk) [inline pkg/timescale/write.go]: Do not call `context.Background()` repeatedly, create one context and use it throughout the function

File: pkg/timescale/write.go [hunk 6]
```diff
@@ -112,35 +147,41 @@ func contains(s []string, e string) bool {
 }
 
 func insertRowsTimescale(rowsArray []TimescaleRows, config *config.OutputTimescale) error {
-	var c, parseErr = pgx.ParseConnectionString(config.Connection)
-	if parseErr != nil {
-		return parseErr
-	}
 
-	conn, connErr := pgx.Connect(c)
+	ctx := context.TODO()
```
Comment (maximiliancsuk) [inline pkg/timescale/write.go]: Is there a reason you are using `context.TODO()` instead of `context.Background`? IMHO, background makes more sense here.

Comment (muhametk) [reply pkg/timescale/write.go]: Since these two are quite similiar I agree that using Background makes more sense.

File: pkg/timescale/write.go [hunk 7]
```diff
@@ -112,35 +147,41 @@ func contains(s []string, e string) bool {
...
-	defer conn.Close()
 
-	tx, beginErr := conn.Begin()
+	defer conn.Release()
+
+	tx, beginErr := conn.Begin(ctx)
 	if beginErr != nil {
 		return beginErr
 	}
-	defer tx.Rollback() //nolint: errcheck
```
Comment (maximiliancsuk) [inline pkg/timescale/write.go:130]: Why did you remove this here and moved the Rollback into the error handling? It was working well before... 

Maybe you were worried about calling Rollbackj on a Comitted transaction? That's ok to do and does not do any harm. See here: https://pkg.go.dev/github.com/jackc/pgx/v4#hdr-Transactions
